content/blog/2022-04-07-compiling-black-with-mypyc-part-1.md

```diff
@@ -0,0 +1,591 @@
+
+### *Phew*, it's faster
+
+It was at this time I finally posted the
+[first set of benchmark numbers][initial-results], they weren't scientific or reliable in
+any way, but it did prove mypyc was still a worthwhile way to improve performance.
+Formatting the Black codebase (so meta, I know) with compiled Black yielded a performance
+gain of 1.82x, impressive for the short amount of time I had put in!
+
+I had to verify this experimental build wasn't completely unstable though. I couldn't just
+run the test suite with compiled Black installed, some tests use mocks which trigger the
+strict type checks mypyc does at runtime. Sadly, I had to mark these tests as straight up
+incompatible so they could be skipped.
+
+### It's still alpha software
+
+As of writing mypyc is still alpha software with rough edges everywhere, I already showed
+quite a few of them above, but there were more 😥 You can read the rest in this
+[integration report][integration-report] I filed in the mypyc issue tracker.
+
+And I know, y'all are wondering why Black didn't use Cython, well I'm not sure as I wasn't
+around in 2019 when [@ambv first suggested its use][ambv-mypyc-suggestion]. Eitherway I
+didn't look into Cython because at the time as it seemed like I only had the final scraps
+left and didn't have too much work to do ... which was both right and wrong depending on
+how you look at it.
+
+Could Black be even faster if we had used Cython? Probably, but that's for another day as
+my personal roadmap for psf/black is already packed :p
+
+**Anyway, it's time for [Pt. 2 - Optimization](../compiling-black-with-mypyc-part-2/).**
+
+[^1]: Originally when I first landed the relevant PR it was an overall 2x improvement, but
+    once Jelle added a stability hotfix the *effective* speedup for files that are changed
+    is 50%. If you're formatting a bunch of already well formatted files, the speedup is
+    still 2x
+
+[^2]: There's too many typing related PEPs to copy and paste, but there's an up to date list
+    here: <https://docs.python.org/3/library/typing.html#relevant-peps>
+
+[^3]: <https://mypyc.readthedocs.io/en/stable/introduction.html#introduction>
+
+[^5]: I am well aware I don't need to import `List` or `Tuple` from `typing` anymore. I'm
+    just keeping my code examples compatible with older versions of Python.
+
+[^4]: well not quite exactly this, it was a crapper solution with a time complexity of O(n²)
+    instead of O(n), oh and of course my code style was less than ideal and I didn't use
+    type annotations, but let's not go there :)
+
+[^6]: At the time @ambv wrote Black, lib2to3 was the best AST + CST mix out there. AST
+    parsers like the built-in one are great for understanding syntax and structure, but
+    are useless when editing how much whitespace a NAME token gets for example. CSTs on
+    the other hand are great for visual changes Black does, but don't provide enough
```

Comment:
I'm actually not sure what the fork was for originally, but I don't think it was for more context. Also I wouldn't call lib2to3 an AST/CST mix; it's very much a CST, which is why it's so annoying to work with.